1.编写 Python程序,判断输入a是否为质数

In [2]:
def is_prime(a):
    if a <= 1:
        return False
    for i in range(2, int(a**0.5) + 1):
        if a % i == 0:
            return False
    return True

# Example
a = 29
print(f"Test: {a} Result:{is_prime(a)}")

Test: 29 Result:True


Test: 29 Result:True

2.用Python实现插入排序该算法，不允许直接调用插入排序函数，需要手动实现其逻辑过程。

In [4]:
def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr

# Example
arr = [12, 11, 13, 5, 6]
sorted_arr = insertion_sort(arr)
print(sorted_arr)

[5, 6, 11, 12, 13]


Output:[5, 6, 11, 12, 13]

3.用Python实现选择排序该算法，不允许直接调用插入排序函数，需要手动实现其逻辑过程。

In [5]:
def selection_sort(arr):
    for i in range(len(arr)):
        min_idx = i
        for j in range(i+1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr

# Example usage
arr = [64, 25, 12, 22, 11]
sorted_arr = selection_sort(arr)
print(sorted_arr)

[11, 12, 22, 25, 64]


Output:[11, 12, 22, 25, 64]

4.用Python实现快速排序该算法，不允许直接调用插入排序函数，需要手动实现其逻辑过程。

In [6]:
def quicksort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quicksort(left) + middle + quicksort(right)

# Example usage
arr = [3, 6, 8, 10, 1, 2, 1]
sorted_arr = quicksort(arr)
print(sorted_arr)

[1, 1, 2, 3, 6, 8, 10]


Output:[1, 1, 2, 3, 6, 8, 10]

5.使用 Python 随机生成三组长度递增的随机数列，长度分别为100、1000、10000，对生成的随机数组使用不同的排序算法，例如选择排序和归并排序，至少三种排序方法。对这些数列的数据排序，并分析不同排序算法在不同长度数列下的运行效果，例如运行时间等指标。

In [7]:
import random
import time

array_100 = [random.randint(0, 1000) for _ in range(100)]
array_1000 = [random.randint(0, 1000) for _ in range(1000)]
array_10000 = [random.randint(0, 1000) for _ in range(10000)]

def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left, right)

def merge(left, right):
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result

def measure_time(sort_func, arr):
    start_time = time.time()
    sort_func(arr.copy())
    end_time = time.time()
    return end_time - start_time

selection_sort_time_100 = measure_time(selection_sort, array_100)
selection_sort_time_1000 = measure_time(selection_sort, array_1000)
selection_sort_time_10000 = measure_time(selection_sort, array_10000)

quicksort_time_100 = measure_time(quicksort, array_100)
quicksort_time_1000 = measure_time(quicksort, array_1000)
quicksort_time_10000 = measure_time(quicksort, array_10000)

merge_sort_time_100 = measure_time(merge_sort, array_100)
merge_sort_time_1000 = measure_time(merge_sort, array_1000)
merge_sort_time_10000 = measure_time(merge_sort, array_10000)

print(f"Selection Sort Time (100): {selection_sort_time_100:.6f} seconds")
print(f"Selection Sort Time (1000): {selection_sort_time_1000:.6f} seconds")
print(f"Selection Sort Time (10000): {selection_sort_time_10000:.6f} seconds")

print(f"Quicksort Time (100: {quicksort_time_100:.6f} seconds")
print(f"Quicksort Time (1000): {quicksort_time_1000:.6f} seconds")
print(f"Quicksort Time (10000: {quicksort_time_10000:.6f} seconds")

print(f"Merge Sort Time (100): {merge_sort_time_100:.6f} seconds")
print(f"Merge Sort Time (1000): {merge_sort_time_1000:.6f} seconds")
print(f"Merge Sort Time (10000): {merge_sort_time_10000:.6f} seconds")

Selection Sort Time (100): 0.000000 seconds
Selection Sort Time (1000): 0.011279 seconds
Selection Sort Time (10000): 1.642781 seconds
Quicksort Time (100: 0.000000 seconds
Quicksort Time (1000): 0.000000 seconds
Quicksort Time (10000: 0.011093 seconds
Merge Sort Time (100): 0.000000 seconds
Merge Sort Time (1000): 0.000000 seconds
Merge Sort Time (10000): 0.016005 seconds


Output:
Selection Sort Time (100): 0.000000 seconds
Selection Sort Time (1000): 0.011279 seconds
Selection Sort Time (10000): 1.642781 seconds
Quicksort Time (100: 0.000000 seconds
Quicksort Time (1000): 0.000000 seconds
Quicksort Time (10000: 0.011093 seconds
Merge Sort Time (100): 0.000000 seconds
Merge Sort Time (1000): 0.000000 seconds
Merge Sort Time (10000): 0.016005 seconds

6.使用python代码获取当前系统时间

In [9]:
from datetime import datetime

current_time = datetime.now()
print(current_time)

2024-10-15 13:06:49.690628


2024-10-15 13:06:49.690628

7.homework/data文件夹下的stuGrade.csv文件中包含5位同学的学号、语文成绩、数学成绩和英语成绩，请先用记事本打开该文件，查看文件内容，然后编写Python程序，读取该文件，并计算出各科的平均成绩（保留2位小数）。

In [16]:
import pandas as pd
df = pd.read_csv('stuGrade.csv')
average_scores = df[['chinese', 'math', 'english']].mean().round(2)
print("Average Scores:")
print(average_scores)
with open('my.txt', 'w') as file:
    file.write("Average Scores:\n")
    file.write(average_scores.to_string())

Average Scores:
chinese    82.6
math       81.0
english    81.8
dtype: float64
Average scores written to my.txt


Average Scores:
chinese    82.6
math       81.0
english    81.8